contrib/core-contract-tests/tests/pox-4/pox_StackStxCommand.ts

```diff
@@ -0,0 +1,173 @@
+import { assert, expect } from "vitest";
+import { Cl, ClarityType, isClarityType } from "@stacks/transactions";
+
+/**
+ * The `StackStxCommand` locks STX for stacking within PoX-4. This self-service
+ * operation allows the `tx-sender` (the `wallet` in this case) to participate
+ * as a Stacker.
+ *
+ * Constraints for running this command include:
+ * - The Stacker cannot currently be engaged in another stacking operation.
+ * - A minimum threshold of uSTX must be met, determined by the
+ *   `get-stacking-minimum` function at the time of this call.
+ * - The amount of uSTX locked may need to be increased in future reward cycles
+ *   if the minimum threshold rises.
+ */
+export class StackStxCommand implements PoxCommand {
+  readonly wallet: Wallet;
+  readonly authId: number;
+  readonly period: number;
+  readonly margin: number;
+
+  /**
+   * Constructs a `StackStxCommand` to lock uSTX for stacking.
+   *
+   * @param wallet - Represents the Stacker's wallet.
+   * @param authId - Unique auth-id for the authorization.
+   * @param period - Number of reward cycles to lock uSTX.
+   * @param margin - Multiplier for minimum required uSTX to stack so that each
+   *                 Stacker locks a different amount of uSTX across test runs.
+   */
+  constructor(
+    wallet: Wallet,
+    authId: number,
+    period: number,
+    margin: number,
+  ) {
+    this.wallet = wallet;
+    this.authId = authId;
+    this.period = period;
+    this.margin = margin;
+  }
+
+  check(model: Readonly<Stub>): boolean {
+    // Constraints for running this command include:
+    // - A minimum threshold of uSTX must be met, determined by the
+    //   `get-stacking-minimum` function at the time of this call.
+    // - The Stacker cannot currently be engaged in another stacking operation.
+    // - The Stacker cannot currently be delegating STX to a delegatee.
+    return (
+      model.stackingMinimum > 0 &&
+      !model.wallets.get(this.wallet.stxAddress)?.isStacking &&
+      !model.wallets.get(this.wallet.stxAddress)?.hasDelegated
+    );
+  }
+
+  run(model: Stub, real: Real): void {
+    // The maximum amount of uSTX that can be used (per tx) with this signer
+    // key. For our tests, we will use the minimum amount of uSTX to be stacked
+    // in the given reward cycle multiplied by the margin, which is a randomly
+    // generated number passed to the constructor of this class.
+    const maxAmount = model.stackingMinimum * this.margin;
+
+    const signerSig = this.wallet.stackingClient.signPoxSignature({
+      // The signer key being authorized.
+      signerPrivateKey: this.wallet.signerPrvKey,
+      // The reward cycle for which the authorization is valid.
+      // For `stack-stx` and `stack-extend`, this refers to the reward cycle
+      // where the transaction is confirmed. For `stack-aggregation-commit`,
+      // this refers to the reward cycle argument in that function.
+      rewardCycle: 0,
```

Comment:
Hmm, maybe misunderstanding but any reason this is a constant?

Reply:
Thank you @setzeus! I think we should turn it into `current pox reward cycle` @moodmosaic.